GPII-3830: Detect if a drive exists before opening it.

stegru opened the pull request:
A better fix than https://github.com/GPII/windows/pull/255

javihernandez commented:
> CI job failed: https://ci.gpii.net/job/windows-tests/936/

```
Unable to determine the bus type for drive P:\: win32 error: DeviceIoControl return:0 win32:0
09:57:56.510:  jq: FAIL: Module "gpii.tests.windows.windowsUtilities" Test name "Testing getUsbDrives" - Message: getUsbDrives should return the expected drive paths
09:57:56.510:  jq: Expected: [
    "G:\\",
    "I:\\",
    "J:\\"
]
09:57:56.511:  jq: Actual: []
09:57:56.511:  jq: Source:     at Object.assertDeepEq (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:180:23)
    at Object.<anonymous> (\\vboxsvr\vagrant\gpii\node_modules\windowsUtilities\test\WindowsUtilitiesTests.js:290:16)
    at Test.run (\\vboxsvr\vagrant\node_modules\infusion\tests\lib\qunit\js\qunit.js:207:18)
    at \\vboxsvr\vagrant\node_modules\infusion\tests\lib\qunit\js\qunit.js:366:10
    at process (\\vboxsvr\vagrant\node_modules\infusion\tests\lib\qunit\js\qunit.js:1458:24)
    at Timeout._onTimeout (\\vboxsvr\vagrant\node_modules\infusion\tests\lib\qunit\js\qunit.js:484:5)
    at ontimeout (timers.js:425:11)
    at tryOnTimeout (timers.js:289:5)
    at listOnTimeout (timers.js:252:5)
    at Timer.processTimers (timers.js:212:10)
09:57:56.512:  jq: Test concluded - Module "gpii.tests.windows.windowsUtilities" Test name "Testing getUsbDrives": 61/62 passed - FAIL
```

I guess you also need to update the failing test, right?

javihernandez commented:
ok to test

stegru commented:
ok to test

stegru commented:
ok to test

```
[K==> default: Matching MAC address for NAT networking...
==> default: Checking if box 'inclusivedesign/windows10-eval-x64-Apps' is up to date...
==> default: Setting the name of the VM: windows-tests_default_1581785900055_11239
==> default: Fixed port collision for 5985 => 5985. Now on port 2249.
==> default: Fixed port collision for 22 => 2222. Now on port 2250.
Vagrant found a port collision for the specified port and virtual machine.
While this port was marked to be auto-corrected, the ports in the
auto-correction range are all also used.

VM: default
Forwarded port: 5985 => 55985
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```

javihernandez commented:
> CI job failed: https://ci.gpii.net/job/windows-tests/1298/

Unrelated failure:

```
09:51:43.010:  jq: FAIL: Module "gpii.tests.windows.processHandling" Test name "Testing closeProcessByName" - Message: Expected 3 assertions, but 2 were run
09:51:43.010:  jq: Source:     at Object.asyncTest (\\vboxsvr\vagrant\node_modules\infusion\tests\lib\qunit\js\qunit.js:406:9)
    at Object.<anonymous> (\\vboxsvr\vagrant\gpii\node_modules\processHandling\test\testProcessHandling.js:165:8)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Module.replacementCompile (\\vboxsvr\vagrant\node_modules\append-transform\index.js:58:13)
    at Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Object.<anonymous> (\\vboxsvr\vagrant\node_modules\append-transform\index.js:62:4)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at Object.<anonymous> (\\vboxsvr\vagrant\tests\UnitTests.js:24:1)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Module.replacementCompile (\\vboxsvr .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
09:51:43.010:  jq: Test concluded - Module "gpii.tests.windows.processHandling" Test name "Testing closeProcessByName": 2/3 passed - FAIL
09:51:43.104:  Executing C:\Users\vagrant\AppData\Local\Temp\gpii-process-handling-test.exe closeProcessByNameTest /T 30
09:51:43.124:  Exit code:5
09:51:43.872:  Process gpii-process-handling-test.exe will not close cleanly: No windows belong to the process
09:51:44.256:  jq: FAIL: Module "gpii.tests.windows.processHandling" Test name "Testing closeProcessByName (window-less process)" - Message: Expected 3 assertions, but 4 were run
09:51:44.256:  jq: Source:     at Object.asyncTest (\\vboxsvr\vagrant\node_modules\infusion\tests\lib\qunit\js\qunit.js:406:9)
    at Object.<anonymous> (\\vboxsvr\vagrant\gpii\node_modules\processHandling\test\testProcessHandling.js:201:8)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Module.replacementCompile (\\vboxsvr\vagrant\node_modules\append-transform\index.js:58:13)
    at Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Object.<anonymous> (\\vboxsvr\vagrant\node_modules\append-transform\index.js:62:4)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at Object.<anonymous> (\\vboxsvr\vagrant\tests\UnitTests.js:24:1)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Module.replacementCompile (\\vboxsvr .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
```
ok to test

javihernandez commented on gpii/node_modules/WindowsUtilities/test/WindowsUtilitiesTests.js:
```diff
@@ -276,6 +286,13 @@ jqUnit.test("Testing getUsbDrives", function () {
         return currentTest.DeviceIoControl;
     };
 
+    var existsSync = fs.existsSync;
```
Ideally, we should make use of contexts, but given that the windows utilities don't belong to any fluid component, I agree that this is the way to go. Still, it would be great if you can add a comment describing the reason for overwriting fs.existsSync. Then I'd be more than happy to merge it.

stegru replied:
done

amb26 commented on gpii/node_modules/WindowsUtilities/test/WindowsUtilitiesTests.js:
```diff
@@ -276,6 +286,13 @@ jqUnit.test("Testing getUsbDrives", function () {
     };
 
+    var existsSync = fs.existsSync;
+    fs.existsSync = function (path) {
```
No, we can't do this - lots of code has broken in the past because of the assumption they can overwrite a core library. Please rewrite a shim for fs.existsSync as gpii.windows.existsSync and redirect all calls to that.

stegru replied:
done

javihernandez commented:
> [<img alt="@stegru" width="20" height="20" src="https://avatars2.githubusercontent.com/u/1867587?s=60&amp;u=90ba9279ea426d59b0845facf75627db8f798491&amp;v=4"> ](/stegru)
> ` [GPII-3830: Mocking a wrapper fs.existsSync, rather than the real func…](/GPII/windows/pull/257/commits/0f250c46386ceb0974531581655898b21bc2414b) ` …

LGTM, @amb26 ?

amb26 commented:
ok to test
